### PR TITLE
ABD-41: Extend setup to also create a virtualenv.

### PR DESCRIPTION
**Pull request by MichaelWalker**
Using virtualenv allows us to
a) have a nice, well isolated python environment in which to develop.
b) package all of our dependencies in a format that we can ship to AWS.

Solo: @michaelwalker

### CODE REVIEW

File: setup.sh [hunk 1]
```diff
@@ -1,2 +1,15 @@
...
\ No newline at end of file
+
+
+if [ -z "$(which python3)" ];
+then
+    echo "install"
+    brew install python3
+else
+    echo "update"
+    brew upgrade python3
```
Comment (kornysietsma) [inline setup.sh]: Hmm - I'm not sure about installing/upgrading python without warnings.  When do people run setup.sh?  I'd prefer something that fails if python is not installed or is too old, with a warning "please run brew install..." - if that's not too much more work.

File: src/event_handler.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        # catch all errors and log them - we never want a single failing message to kill the process.
+        try:
+            event = event_from_json(message['Body'])
+            print(event.event_id)
+            __delete_message(sqs_client, queue_url, message)
+        except Exception as exception:
+            print('Failed to store message - Exception: {0}'.format(str(message)))
+
+
+def __fetch_single_message(sqs_client, queue_url):
```
Comment (kornysietsma) [inline src/event_handler.py]: More of a question than a comment - when does SQS mark a message as consumed?  It's important for our event to be sort-of transactional.  If we receive a message, then start writing it to the database, and then the database write fails, or the process crashes, or lightning hits the data center, I'd want the SQS message to stay on the queue and be (eventually) re-delivered.

File: test/event_handler_test.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+        event_handler.store_queued_events(None, None)
+
+        self.assertEqual(mock_stdout.getvalue(), 'sample-id-1\nsample-id-2\n')
+        self.assertEqual(self.__number_of_visible_messages(), '0')
+        self.assertEqual(self.__number_of_hidden_messages(), '0')
+
+    @patch('sys.stdout', new_callable=io.StringIO)
+    def test_does_not_delete_invalid_messages(self, mock_stdout):
+        self.__send_messages_to_queue(
```
Comment (kornysietsma) [inline test/event_handler_test.py]: Something to discuss - should we be moving messages to a dead letter queue?  Does SQS do this somehow automatically?  (not a blocker for this PR!)

File: src/event_handler.py (hunk 2 above)
Comment (MichaelWalker) [reply src/event_handler.py]: As I understand it, SQS won't mark it as consumed until we explicitly 'delete' the message from the queue. 
Deleting the message is the last thing we do, so any exception would mean that the message remains on the queue.

We have a test for this. When the message is invalid, an error is thrown, and we assert that there should be one (not visible) message remaining in the queue.

File: test/event_handler_test.py (hunk 3 above)
Comment (MichaelWalker) [reply test/event_handler_test.py]: Yes, SQS does this for us, which is a nice separation of concerns.

SQS keeps track of how many times a message has become visible, and you can configure SQS with a 'max retry count' (actual name might be slightly different). You can configure behaviour on reaching this count, including sending to DLQ

File: setup.sh (hunk 1 above)
Comment (MichaelWalker) [reply setup.sh]: Yeah, Vanessa and I spent a while chatting about this, and didn't come up with anything that we thought was perfect.

I prefer your suggestion though, so I'll change it to that in a new PR.

**Pull request merged**